zip-0221.rst

```diff
@@ -0,0 +1,743 @@
+in the right subtree of the MMR root.
+
+.. code-block:: python
+
+    def delete(root: ZcashMMRNode) -> ZcashMMRNode:
+        '''
+        Delete the rightmost leaf node from an existing MMR
+        Return the new tree root
+        '''
+
+        n_leaves = root.latest_height - (root.earliest_height - 1)
+        # if there were an odd number of leaves,
+        # simply replace root with left_child
+        if n_leaves & 1:
+            return root.left_child
+
+        # otherwise, we need to re-bag the peaks.
+        else:
+            # first peak
+            peaks = [root.left_child]
+
+            # we do this traversing the right (unbalanced) side of the tree
+            # we keep the left side (balanced subtree or leaf) of each subtree
+            # until we reach a leaf
+            subtree_root = root.right_child
+            while subtree_root.left_child:
+                peaks.push(subtree_root.left_child)
+                subtree_root = subtree_root.right_child
+
+        new_root = bag_peaks(peaks)
+        return new_root
+
+Block header semantics and consensus rules
+------------------------------------------
+
+The ``hashFinalSaplingRoot`` block header field (which was named ``hashReserved`` prior to
+the Sapling network upgrade) is renamed to ``hashLightClientRoot``, to reflect its usage
+by light clients.
+
+Prior to activation of the Heartwood network upgrade, this existing consensus rule on
+block headers (adjusted for the renamed field) is enforced: [#block-header]_
+
+    [Sapling onward] ``hashLightClientRoot`` MUST be LEBS2OSP<sub>256</sub>(rt) where rt
+    is the root of the Sapling note commitment tree for the final Sapling tree state of
+    this block.
+
+Once the Heartwood network upgrade activates, ``hashLightClientRoot`` MUST be set to the
+value of ``hashChainHistoryRoot`` as specified above.
+
+The block header byte format and version are not altered by this ZIP.
+
+
+Rationale
+=========
+
+Tree nodes
+----------
+
+Nodes in the commitment tree are canonical and immutable. They are cheap to generate, as
+(with the exception of ``nSaplingTxCount``) all metadata is already generated during
+block construction and/or checked during block validation. Nodes are relatively compact in
+memory. Approximately 140,000 blocks have elapsed since Sapling activation. Assuming a 164
+byte commitment to each of these, we would have generated approximately 24 MB of
+additional storage cost for the set of leaf nodes (and an additional ~24 MB for storage of
+intermediate nodes).
+
+``hashSubtreeCommitment`` forms the strucuture of the commitment tree. Other metadata
+commitments were chosen to serve specific purposes. Variable-length commitments are placed
+last, so that most metadata in a node can be directly indexed. We considered using
+fixed-length commitments here, but opted for variable-length, in order to marginally
+reduce the memory requirements for managing and updating the commitment trees.
```

Comment:
The CompactSize fields are likely to be at least 2 bytes each on average, so this only saves ~18 bytes per commitment out of 176 bytes (assuming the length fields are changed to 64-bit), i.e. ~10%.